### PR TITLE
Tipline improvements and fixes

### PR DESCRIPTION
**Pull request by caiosba**
## Description

This branch implements a bunch of small improvements and fixes for tiplines, specially for the newsletter feature.

**Output more clear error message in log when newsletter can't be sent to a subscriber** (CV2-3372)
Make sure that all tipline providers (Smooch, On-Prem and CAPI) return a response object with attributes `code` and `body` for the `send_message_to_user` method. In the tipline newsletter delivery job, log successful deliveries and failed deliveries according to the `response.code`.

**When sending a newsletter, be sure that RSS content is not re-fetched for each subscriber** (CV2-3371)
"Cache" newsletter articles in memory, so, for the same `TiplineNewsletter` object loaded in memory, the method `articles` won't re-fetch static articles or articles from an RSS feed after the first time this method is called. This is for performance reasons... when a newsletter is delivered, that method is called several times (once for each subscription), so RSS newsletters can take a long time if there are many subscribers and the RSS URL doesn't load fast enough.

**Shorten URLs in newsletter introduction** (CV2-3259)
Pretty straightforward, just need to call a method around the newsletter introduction to shorten the URLs.

**Set event `newsletter` for `TiplineMessage` records related to newsletter deliveries** (CV2-3298)
Save the sent message ID so that the "newsletter" delivery event can be stored in the `TiplineMessage`.

**Get caption for media files sent to on-prem tiplines** (CV2-3288)
This is bug fixes. Affects only WhatsApp tiplines currently running on the on-prem API. The fix is to simply reuse the same incoming message parser used for tiplines running in the WhatsApp Cloud API.

**Avoid edge cases where a message is misinterpreted as a language code** (CV2-3312)
For example, "hi" misinterpreted as "Hindi".

## How has this been tested?

Unit tests for all cases. 100% coverage should be maintained.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [x] I have added regression tests, if the PR fixes a bug
- [x] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [x] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [x] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

File: app/models/bot/smooch.rb [hunk 1]
```diff
@@ -430,7 +430,7 @@ def self.get_custom_menu_options(state, workflow, uid)
         self.get_supported_languages.each do |l|
           i = self.get_next_menu_item_number(i)
           options << {
-            'smooch_menu_option_keyword' => [l, i].join(','),
+            'smooch_menu_option_keyword' => i.to_s,
```
Comment (caiosba) [inline app/models/bot/smooch.rb:433]: Do not add the language code as a valid key for a tipline menu language option. Now the key is just a number, which is the order of that option in the menu. This was supposed to serve as shortcuts for language codes but there are edge cases (for example, "hi" can be misinterpreted as a shortcut for "Hindi", if the tipline supports that language).

File: app/models/bot/smooch.rb [hunk 2]
```diff
@@ -935,10 +935,14 @@ def self.get_id_from_send_response(response)
...
-  def self.save_smooch_response(response, pm, query_date, fallback_template = nil, lang = 'en', custom = {})
+  def self.save_smooch_response(response, pm, query_date, fallback_template = nil, lang = 'en', custom = {}, expire = nil)
     return false if response.nil? || fallback_template.nil?
     id = self.get_id_from_send_response(response)
-    Rails.cache.write('smooch:original:' + id, { project_media_id: pm&.id, fallback_template: fallback_template, language: lang, query_date: (query_date || Time.now.to_i) }.merge(custom).to_json) unless id.blank?
+    unless id.blank?
+      key = 'smooch:original:' + id
+      value = { project_media_id: pm&.id, fallback_template: fallback_template, language: lang, query_date: (query_date || Time.now.to_i) }.merge(custom).to_json
+      Rails.cache.write(key, value, expires_in: expire)
+    end
```
Comment (caiosba) [inline app/models/bot/smooch.rb:945]: Make sure that these Redis keys can expire, otherwise Redis space usage can grow indefinitely.

File: app/models/concerns/smooch_resend.rb [hunk 3]
```diff
@@ -219,16 +219,17 @@ def template_locale_options(team_slug = nil)
     end
 
     def format_template_message(template_name, placeholders, file_url, fallback, language, file_type = 'image', preview_url = true)
+      provider = RequestStore.store[:smooch_bot_provider]
       namespace = self.config['smooch_template_namespace']
       template = self.config["smooch_template_name_for_#{template_name}"] || template_name
-      return '' if namespace.blank? || template.blank?
+      return '' if ['TURN', 'CAPI'].include?(provider) && (namespace.blank? || template.blank?)
```
Comment (caiosba) [inline app/models/concerns/smooch_resend.rb:225]: Templates are only for WhatsApp. So this verification is to be sure that execution can continue and the message can be sent for non-WhatsApp tiplines even if there is no template namespace and template name defined.

File: app/models/concerns/smooch_turnio.rb [hunk 4]
```diff
@@ -97,7 +97,7 @@ def get_turnio_message_event(json)
     end
 
     def get_turnio_message_text(message)
-      message.dig('text', 'body') || message.dig('interactive', 'list_reply', 'title') || message.dig('interactive', 'button_reply', 'title') || ''
+      Bot::Smooch.get_capi_message_text(message)
```
Comment (caiosba) [inline app/models/concerns/smooch_turnio.rb:100]: WhatsApp Cloud API and WhatsApp On-Prem API have the same payload for sent messages. So, avoid code duplication by using only one method for both.

File: app/models/concerns/smooch_turnio.rb [hunk 5]
```diff
@@ -273,7 +270,7 @@ def turnio_send_message_to_user(uid, text, extra = {}, _force = false)
           )
         end
       end
-      ret
+      response
```
Comment (caiosba) [inline app/models/concerns/smooch_turnio.rb:273]: Always return a HTTP response, regardless if it failed or succeeded.

File: app/models/concerns/smooch_zendesk.rb [hunk 6]
```diff
@@ -69,12 +69,16 @@ def zendesk_send_message_to_user(uid, text, extra = {}, _force = false)
           }
         })
       end
-      return if params['type'] == 'text' && params['text'].blank?
+      return OpenStruct.new(body: 'Empty message', code: 400) if params['type'] == 'text' && params['text'].blank?
```
Comment (caiosba) [inline app/models/concerns/smooch_zendesk.rb:72]: Return an object that looks like a HTTP response (e.g., contains code and body).

File: app/models/tipline_newsletter.rb [hunk 7]
```diff
@@ -134,6 +134,7 @@ def whatsapp_template_name
   end
 
   def articles
+    return @articles unless @articles.to_a.empty?
```
Comment (caiosba) [inline app/models/tipline_newsletter.rb:137]: Avoid re-fetching articles from RSS feed if this same method is called many times for the same object already in memory. It's like a cheap cache.

**Comment by caiosba** [comment]
Ready for review, @melsawy @hartsick . I added comments to the code explaining the idea of each change.

**Review by melsawy — APPROVED** [review]
